tests/azure_iot_cmocka_mqtt.c

```diff
@@ -0,0 +1,142 @@
+/* SPDX-License-Identifier: MIT */
+
+/**
+ * @file azure_iot_cmocka_mqtt.c
+ * @brief Unit test dummy MQTT port.
+ *
+ */
+
+#include <stdarg.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <setjmp.h>
+
+#include <cmocka.h>
+
+#include "azure_iot_mqtt.h"
+#include "azure_iot_mqtt_port.h"
+
+AzureIoTMQTTEventCallback_t xTestUserCallback = NULL;
+AzureIoTMQTTPacketInfo_t xPacketInfo;
+AzureIoTMQTTDeserializedInfo_t xDeserializedInfo;
+uint16_t usTestPacketId = 1;
+const uint8_t * pucPublishPayload = NULL;
+
+AzureIoTMQTTResult_t AzureIoTMQTT_Init( AzureIoTMQTTHandle_t xContext,
+                                        const AzureIoTTransportInterface_t * pxTransportInterface,
+                                        AzureIoTMQTTGetCurrentTimeFunc_t xGetTimeFunction,
+                                        AzureIoTMQTTEventCallback_t xUserCallback,
+                                        uint8_t * pucNetworkBuffer,
+                                        size_t xNetworkBufferLength )
+{
+    ( void ) xContext;
+    ( void ) pxTransportInterface;
+    ( void ) xGetTimeFunction;
+    ( void ) pucNetworkBuffer;
+    ( void ) xNetworkBufferLength;
+
+    xTestUserCallback = xUserCallback;
+    return ( AzureIoTMQTTResult_t ) mock();
+}
+
+AzureIoTMQTTResult_t AzureIoTMQTT_Connect( AzureIoTMQTTHandle_t xContext,
+                                           const AzureIoTMQTTConnectInfo_t * pxConnectInfo,
+                                           const AzureIoTMQTTPublishInfo_t * pxWillInfo,
+                                           uint32_t ulMilliseconds,
+                                           uint8_t * pucSessionPresent )
+{
+    ( void ) xContext;
+    ( void ) pxConnectInfo;
+    ( void ) pxWillInfo;
+    ( void ) ulMilliseconds;
+    ( void ) pucSessionPresent;
+
+    return ( AzureIoTMQTTResult_t ) mock();
+}
+
+AzureIoTMQTTResult_t AzureIoTMQTT_Subscribe( AzureIoTMQTTHandle_t xContext,
+                                             const AzureIoTMQTTSubscribeInfo_t * pxSubscriptionList,
+                                             size_t xSubscriptionCount,
+                                             uint16_t usPacketId )
+{
+    ( void ) xContext;
+    ( void ) pxSubscriptionList;
+    ( void ) xSubscriptionCount;
+    ( void ) usPacketId;
+
+    return ( AzureIoTMQTTResult_t ) mock();
+}
+
+AzureIoTMQTTResult_t AzureIoTMQTT_Publish( AzureIoTMQTTHandle_t xContext,
+                                           const AzureIoTMQTTPublishInfo_t * pxPublishInfo,
+                                           uint16_t usPacketId )
+{
+    ( void ) xContext;
+    ( void ) usPacketId;
+
+    AzureIoTMQTTResult_t xReturn = ( AzureIoTMQTTResult_t ) mock();
+
+    if( xReturn )
+    {
+        return xReturn;
+    }
+
+    if( pucPublishPayload )
+    {
+        assert_memory_equal( pxPublishInfo->pvPayload, pucPublishPayload, pxPublishInfo->xPayloadLength );
```

Comment:
Love being able to check the outgoing payload.  Do we use this anywhere in the actual test - I'm not seeing `pucPublishPayload` referenced in the UT except in the `external`.

Reply:
I was using these in iothub tests.